Objects/lazyimportobject.c

```diff
@@ -0,0 +1,212 @@
+
+/* Lazy object implementation */
+
+#include "Python.h"
+#include "pycore_lazyimport.h"
+
+PyObject *
+_PyLazyImport_NewModule(
+    PyObject *name, PyObject *globals, PyObject *locals, PyObject *fromlist, PyObject *level)
+{
+    PyLazyImportObject *m;
+    if (!name || !PyUnicode_Check(name) ||
+        !globals || !locals || !fromlist) {
+        PyErr_BadArgument();
+        return NULL;
+    }
+    if (level == NULL) {
+        level = PyLong_FromLong(0);
+        if (level == NULL) {
+            return NULL;
+        }
+    } else {
+        Py_INCREF(level);
+    }
+    m = PyObject_GC_New(PyLazyImportObject, &PyLazyImport_Type);
+    if (m == NULL) {
+        Py_DECREF(level);
+        return NULL;
+    }
+    m->lz_lazy_import = NULL;
+    Py_INCREF(name);
+    m->lz_name = name;
+    Py_INCREF(globals);
+    m->lz_globals = globals;
+    Py_INCREF(locals);
+    m->lz_locals = locals;
+    Py_INCREF(fromlist);
+    m->lz_fromlist = fromlist;
+    m->lz_level = level;
+    m->lz_resolved = NULL;
+    m->lz_resolving = NULL;
+    PyObject_GC_Track(m);
+    return (PyObject *)m;
+}
+
+PyObject *
+_PyLazyImport_NewObject(PyObject *from, PyObject *name)
+{
+    PyLazyImportObject *m;
+    if (!from || !PyLazyImport_CheckExact(from) || !name || !PyUnicode_Check(name)) {
+        PyErr_BadArgument();
+        return NULL;
+    }
+    m = PyObject_GC_New(PyLazyImportObject, &PyLazyImport_Type);
+    if (m == NULL) {
+        return NULL;
+    }
+    PyLazyImportObject *lazy_from = (PyLazyImportObject *)from;
+    if (lazy_from->lz_fromlist != NULL && lazy_from->lz_fromlist != Py_None) {
+        PyObject *fromlist = PyTuple_New(1);
+        if (fromlist == NULL) {
+            return NULL;
+        }
+        Py_INCREF(name);
+        if (PyTuple_SetItem(fromlist, 0, name) < 0) {
+            return NULL;
+        }
+        PyObject *new_from = _PyLazyImport_NewModule(
+            lazy_from->lz_name, lazy_from->lz_globals, lazy_from->lz_locals, fromlist, lazy_from->lz_level);
+        Py_DECREF(fromlist);
+        if (new_from == NULL) {
+            return NULL;
+        }
+        m->lz_lazy_import = new_from;
+    } else {
+        Py_INCREF(from);
+        m->lz_lazy_import = from;
+    }
```

Comment:
I may not fully understand how `lz_fromlist` works. Why would we have a `from` with empty `lz_fromlist`? If that is possible, is it also possible that our `from` would have a non-empty `lz_fromlist` that doesn't include us? Why don't we need to check for that too?

Reply:
The only possible way to have a from with an empty list is, I believe, with the example shown in the previous question:

```py
import foo.bar.baz.qux as Q
```

In that case it acts as a regular import (no `fromlist`), yet it uses `IMPORT_FROM` to extract the whole path (`"bar"`, `"baz"`, `"qux"`) into `Q`.